netaddr.go

```diff
@@ -412,96 +323,82 @@ func (ip IP) Is6() bool {
 // returns the wrapped IPv4 address. Otherwise it returns ip, regardless
 // of its type.
 func (ip IP) Unmap() IP {
-	if !ip.Is4in6() {
-		return ip
+	if ip.Is4in6() {
+		ip.z = z4
 	}
-	a := ip.ipImpl.as16()
-	return IP{v4Addr{a[12], a[13], a[14], a[15]}}
+	return ip
 }
 
 // WithZone returns an IP that's the same as ip but with the provided
-// zone. If zone is empty, the zone is removed. If ip is an IPv4
+// zone. If zoneName is empty, the zone is removed. If ip is an IPv4
 // address it's returned unchanged.
-func (ip IP) WithZone(zone string) IP {
-	if zone == "" {
-		if z, ok := ip.ipImpl.(v6AddrZone); ok {
-			return IP{z.v6Addr}
-		}
+func (ip IP) WithZone(zoneName string) IP {
+	if !ip.Is6() {
 		return ip
 	}
-	switch ip := ip.ipImpl.(type) {
-	case v6Addr:
-		return IP{v6AddrZone{ip, zone}}
-	case v6AddrZone:
-		return IP{v6AddrZone{ip.v6Addr, zone}}
+	if zoneName == "" {
```

Comment:
We could do `z6noz = intern.Get("")` as the initializer, at which point we could remove this check.

Reply:
Done.